### ISSUE TITLE
Completion when there's only one candidate, and that candidate is already in the buffer?

### ISSUE THREAD

**Issue by cpitclaudel** [company-mode/company-mode#451]
Hey Dmitry,

I have three use cases in company-coq where it would be useful to not disable completion when there's a single candidate and that candidate matches the contents of the buffer. The bit of code I'm talking about is this one in `company-calculate-candidates`:

 ```
      (if (or (cdr candidates)
              (not (eq t (compare-strings (car candidates) nil nil
                                          prefix nil nil ignore-case))))
          candidates
        ;; Already completed and unique; don't start.
        t))))
```

I agree it generally makes sense, but I have the following scenarios:

* There's a post-action associated with that single candidate, and the user wants to run it. Right now they have to delete the last letter of the prefix and call `-manual-begin`
* There's documentation or location information associated with the candidate, and the user wants to see it. company is really good at accessing documentation, and it makes sense for my users to remember a single shortcut to show docs. Same for the metadata: sometimes I just want to see it again for the symbol I just inserted, and so I go back, erase one character, and run `-manual-begin`.
* I want to know whether an identifier is defined (assuming all defined symbols are available for completion). At the moment, I need to delete one character before running `-manual-begin`.

In these three cases, getting a single obvious completion would be very useful :)

It has two issues, however:

*  Not all frontends know how to display a single candidate, especially if there's nothing left to insert;t that's in particular true of the inline ones. To fix this, they could be displayed as if a thin space was 
going to be inserted after the prefix.
* `RET` annoyingly gets rebound to a no-op. That is, pressing ret will "complete" an already complete word; that's annoying, but it's nothing that can't be fixed in a post-command. In fact I already do that in one of company-coq's backends:
```elisp
(defun company-coq-reserved-keywords-backend (command &optional arg &rest ignored)
  "`company-mode' backend for reserved keywords.
COMMAND, ARG and IGNORED: see `company-backends'.
This is mostly useful to prevent completion from kicking in
instead of inserting a newline, after e.g. typing [with]."
  (interactive (list 'interactive))
  (company-coq-dbg "reserved keywords backend: called with command %s" command)
  (pcase command
    (`interactive (company-begin-backend 'company-coq-reserved-keywords-backend))
    (`prefix (company-coq-prefix-at-point))
    (`candidates (company-coq-candidates-reserved arg))
    (`post-completion (call-interactively #'newline))
    (`annotation "<reserved>")))
```
This backend has a high priority in a large multi-backend (about 15 sources), and essentially ensures that pressing `RET` after typing a full keyword (of the underlying programming language) does in fact insert a newline, instead of completing to nothing. It's a very limited solution to the problem that if you have fuzzy matches (or simply duplicate candidates) then typing a candidate in full and pressing enter still only "inserts" that candidate, instead of inserting a newline. Removing the special case of "one candidate, nothing left to complete" would accentuate this problem, but already exists.

Do you have thoughts about this? It's not even clear to me at which semantic level this would happen. Backends? (so a backend could say "my candidates are really important, always show them")?

Thanks!

**Comment by dgutov** [comment]
Indeed, it's a long-standing question. We've encountered it before in #150 and #205. I'd be happy to see it resolved.

The issue with the preview frontend not showing anything for a single fully typed completion is one of the pitfalls. `RET` being no-op is another. But as long as completions have annotations, and `post-completion` performs snippet expansion (this is the case for both `company-clang` and `company-yasnippet`), both issues are solvable.

However, for backends that don't do much in `post-completion`, the other needs can be solved in a different way:

* To show documentation or location of an already-typed completion, just create two new commands and bind them in `company-coq-map`. This shouldn't take much code. And as a bonus (though I don't know if it applies for Coq), you'll be able to resolve ambiguous identifiers based on the contents of the buffer after point. Company itself couldn't do that (and completion candidates shouldn't, in general, depend on contents after point), so language-specific commands can be more precise.

* "I want to know whether an identifier is defined"--define a Flycheck checker? I imagine you'd need one anyway. It will check the defined-ness of identifiers in the whole file, not just the one at point.

> (`post-completion (call-interactively #'newline))

FWIW, I don't particularly like this solution. If users expect `RET` to always insert a newline, maybe avoid using it for completion? Binding `TAB` to something handy should suffice.

> a backend could say "my candidates are really important, always show them"

As per above, I've only considered this issue in the context of backends that perform snippet expansion. To make it clearer to the framework which backends do, we'd introduce a new backend command: `snippet`, which would return a string in Yasnippet format (`company-template` would have to support it as well).

And if the backend returns a non-nil snippet, we'd stop before simply exiting completion. Although exact details are up for discussion: e.g. we could perform snippet expansion right away (and *then* exit completion), instead of waiting for `RET`.

**Comment by cpitclaudel** [comment]
> Indeed, it's a long-standing question. We've encountered it before in #150 and #205. I'd be happy to see it resolved.

Ah, thanks for the pointers!

> The issue with the preview frontend not showing anything for a single fully typed completion is one of the pitfalls. RET being no-op is another. But as long as completions have annotations, and post-completion performs snippet expansion (this is the case for both company-clang and company-yasnippet), both issues are solvable.
> 
> However, for backends that don't do much in post-completion, the other needs can be solved in a different way:
>
> To show documentation or location of an already-typed completion, just create two new commands and bind them in company-coq-map. This shouldn't take much code. And as a bonus (though I don't know if it applies for Coq), you'll be able to resolve ambiguous identifiers based on the contents of the buffer after point. Company itself couldn't do that (and completion candidates shouldn't, in general, depend on contents after point), so language-specific commands can be more precise.

Done as of today :) In https://github.com/cpitclaudel/company-coq/commit/c400ab93eb5a66abae0e61f743d6a51ead82692d and https://github.com/cpitclaudel/company-coq/commit/544a2dee99c8793559e414b86e91060139135041

> "I want to know whether an identifier is defined"--define a Flycheck checker? I imagine you'd need one anyway. It will check the defined-ness of identifiers in the whole file, not just the one at point.

I really really wish we could use Flycheck with Coq, but Coq is just to slow. Processing an entire file can take hours. The model here is you have a process in the background that you send portions of the buffer one by one, incrementally.

>> (`post-completion (call-interactively #'newline))
>>
> FWIW, I don't particularly like this solution. If users expect RET to always insert a newline, maybe avoid using it for completion? Binding TAB to something handy should suffice.

That's fair. The problem is, a quick sampling of the users of company-coq suggests that not many of them use many other Emacs features (in fact, for a while, Proof General was distributed as a customized build of Emacs), so requiring them to rebind keys is not ideal.

>> a backend could say "my candidates are really important, always show them"

> As per above, I've only considered this issue in the context of backends that perform snippet expansion. To make it clearer to the framework which backends do, we'd introduce a new backend command: snippet, which would return a string in Yasnippet format (company-template would have to support it as well).

> And if the backend returns a non-nil snippet, we'd stop before simply exiting completion. Although exact details are up for discussion: e.g. we could perform snippet expansion right away (and then exit completion), instead of waiting for RET.

Fair enough; though I still feel that it would be nice to have an option to see candidates even when there's only one :) Maybe that could be a decision that the frontends take?

**Comment by dgutov** [comment]
> Done as of today :)

:thumbsup:

> The model here is you have a process in the background that you send portions of the buffer one by one, incrementally.

I suppose you could have a command that will verify the current line, then. This problem is pretty far from my daily experience to say what solution is best.

> Fair enough; though I still feel that it would be nice to have an option to see candidates even when there's only one :)

Better suggestions welcome, of course.

> Maybe that could be a decision that the frontends take?

Not the frontends: if `company-calculate-candidates` returns nil upon seeing the "completed and unique" candidate, it will be too late to do anything in frontends.

We'll need some sane model to decide when not to exit completion in this case. We probably can't do that for just any backend: if the sole candidate is typed out, and there's no snippet expansion in sight (or anything similarly useful), the user might be miffed if `company-active-map` is still hogging some useful key bindings. Especially those users that bind `C-n` and `C-p` in it.

**Comment by cpitclaudel** [comment]
> We probably can't do that for just any backend: if the sole candidate is typed out, and there's no snippet expansion in sight (or anything similarly useful), the user might be miffed if company-active-map is still hogging some useful key bindings. Especially those users that bind C-n and C-p in it.

Very good point :)

**Comment by dgutov** [comment]
I think we've satisfied every item on that list now. If not, please let me know.
